### PR TITLE
Support resource group name override in remove test resources template

### LINKED ISSUE #1904: resource deployment doesn't clean-up correctly when there are multiple deployments
Based on pipelines like the spring one, see error https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1047169&view=logs&j=6ddc3820-470e-5638-11a6-c1adaee6c311&t=2fb17bf8-5823-5206-f5c7-887b3d5da607. Ignore the specific error in that log that is handled separately but the main point of this issue is the fact that we call remove resources multiple times with the same resource group. Most likely the resource group we deployed last. We need to come up with a way to ensure we clean-up all the resource groups we deployed if we deploy more then one in a pipeline.

### PR DESCRIPTION
**Pull request by benbp**
Fixes #1904

This changes the remove resources script to stop referencing the static `AZURE_RESOURCEGROUP_NAME` variable and instead reference the `<SERVICEDIRECTORY>_RESOURCE_GROUP` variable, which can support scenarios where multiple groups are deployed in a pipeline.

### CODE REVIEW

**Review by heaths — APPROVED** [review]

**Comment by benbp** [comment]
@heaths sorry I added an update from the prior approach. I found a couple reasons why it wouldn't work in the pipelines.

**Review by mikeharder — APPROVED** [review]

File: eng/common/TestResources/SubConfig-Helpers.ps1 [hunk 1]
```diff
@@ -2,6 +2,17 @@ function BuildServiceDirectoryPrefix([string]$serviceName) {
...
+# If the ServiceDirectory has multiple segments use the last directory name
+# e.g. D:\foo\bar -> bar or foo/bar -> bar
+function GetServiceName([string]$serviceDirectory) {
+    $serviceName = if (Split-Path $serviceDirectory) {
+        Split-Path -Leaf $serviceDirectory
+    } else {
+        $serviceDirectory.Trim('/')
+    }
+    return $serviceName
+}
```
Comment (heaths) [inline eng/common/TestResources/SubConfig-Helpers.ps1:14]: Tip: Just returning `Split-Path -Leaf $ServiceDirectory` should be enough e.g.,

```powershell
PS> @(
  'D:\sdk\keyvault'
  'sdk/keyvault'
  'sdk/keyvault/Azure.Security.KeyVault.Keys'
  'D:\sdk\keyvault\Azure.Security.KeyVault.Keys'
  'sdk/keyvault/'
  'keyvault'
  'keyvault/'
) | Split-Path -Leaf

keyvault
keyvault
Azure.Security.KeyVault.Keys
Azure.Security.KeyVault.Keys
keyvault
keyvault
keyvault
```

File: eng/common/TestResources/Remove-TestResources.ps1 [hunk 2]
```diff
@@ -126,18 +131,27 @@ if ($ProvisionerApplicationId) {
...
+        }
+    } else {
+        # Make sure $BaseName is set.
+        if (!$BaseName) {
+            $UserName = if ($env:USER) { $env:USER } else { "${env:USERNAME}" }
+            # Remove spaces, etc. that may be in $UserName
+            $UserName = $UserName -replace '\W'
+
+            $BaseName = "$UserName$ServiceDirectory"
+            Log "BaseName was not set. Using default base name '$BaseName'"
```
Comment (heaths) [inline eng/common/TestResources/Remove-TestResources.ps1:149]: Nit: Given we do this in New-TestResoruces.ps1, I wonder if this should be moved to the helpers.

**Review by heaths — APPROVED** [review]

File: eng/common/TestResources/Remove-TestResources.ps1 (hunk 2 above)
Comment (benbp) [reply eng/common/TestResources/Remove-TestResources.ps1:149]: Good idea.

File: eng/common/TestResources/SubConfig-Helpers.ps1 [hunk 3]
```diff
@@ -2,6 +2,12 @@ function BuildServiceDirectoryPrefix([string]$serviceName) {
     return $serviceName.ToUpperInvariant() + "_"
 }
 
+# If the ServiceDirectory has multiple segments use the last directory name
+# e.g. D:\foo\bar -> bar or foo/bar -> bar
+function GetServiceName([string]$serviceDirectory) {
+    return Split-Path -Leaf $serviceDirectory
```
Comment (weshaggard) [inline eng/common/TestResources/SubConfig-Helpers.ps1:8]: Is this going to be unique enough? Do we still want to trim?

File: eng/common/TestResources/New-TestResources.ps1 [hunk 4]
```diff
@@ -553,16 +547,12 @@ try {
...
-            $EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'] -ne $ResourceGroupName)
-        {
-            Write-Warning ("Overwriting 'EnvironmentVariables.AZURE_RESOURCEGROUP_NAME' with value " +
-                "'$($EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'])' " + "to new value '$($ResourceGroupName)'")
-        }
-        $EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'] = $ResourceGroupName
+        # Set an environment variable marking that resources have been deployed
+        # This variable can be consumed as a yaml condition in later stages of the pipeline
+        # to determine whether resources should be removed.
+        Write-Host "Setting variable 'CI_HAS_DEPLOYED_RESOURCES': 'true'"
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:553]: Are we using this anywhere yet?

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:548]: I replaced the previous condition in the `remove-test-resources.yml` template with this. https://github.com/Azure/azure-sdk-tools/blob/9dac973bd3b1c5868b5fc8818211ef22bdbb445e/eng/common/TestResources/remove-test-resources.yml#L37

I believe this conditional is there to handle cases where no ARM template is deployed, so wanted to cover this possible backwards compatibility/edge case.

File: eng/common/TestResources/SubConfig-Helpers.ps1 (hunk 3 above)
Comment (benbp) [reply eng/common/TestResources/SubConfig-Helpers.ps1:8]: I was thinking that due to the new multi-segment service directories, however I like the brevity/simplicity/consistency of keeping the leaf. Having to reference environment variable names like `MESSAGING_AZSERVICEBUS` is going to get tedious quickly. Uniqueness is not a concern because:

1. The service name value, when not combined with other unique values, is only used for environment variable prefixes.
2. When the service name is included in the resource group name, it will either contain the local username or a random short string for CI.
    - This could be slightly confusing with groups named like `rg-internal-tf234asd45` from go, but I think the tradeoff for short environment variables is worth it. Plus we can always reverse lookup resource groups by service with the BuildId tag (or by looking at the deployed resource types).

Comment (weshaggard) [reply eng/common/TestResources/SubConfig-Helpers.ps1:8]: > The service name value, when not combined with other unique values, is only used for environment variable prefixes.

Might be worth naming the function as such so there isn't any confusion of if someone else tries use it expecting a service name.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 5]
```diff
@@ -553,16 +542,12 @@ try {
             BuildReason = "${env:BUILD_REASON}"
         }
 
-        # Set the resource group name variable.
-        Write-Host "Setting variable 'AZURE_RESOURCEGROUP_NAME': $ResourceGroupName"
-        LogVsoCommand "##vso[task.setvariable variable=AZURE_RESOURCEGROUP_NAME;]$ResourceGroupName"
-        if ($EnvironmentVariables.ContainsKey('AZURE_RESOURCEGROUP_NAME') -and `
-            $EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'] -ne $ResourceGroupName)
-        {
-            Write-Warning ("Overwriting 'EnvironmentVariables.AZURE_RESOURCEGROUP_NAME' with value " +
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:562]: I guess we were ignoring this warning in our use case :(

File: eng/common/TestResources/New-TestResources.ps1 [hunk 6]
```diff
@@ -553,16 +542,12 @@ try {
...
-                "'$($EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'])' " + "to new value '$($ResourceGroupName)'")
-        }
-        $EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'] = $ResourceGroupName
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:565]: Did we consider just turning the existing ENV into a list?

**Review by weshaggard — APPROVED** [review]
Left a few comments/questions but looks reasonable otherwise.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 5 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:562]: We actually weren't hitting this warning at all, because we run the equality check _after_ we set the variable above...

File: eng/common/TestResources/New-TestResources.ps1 (hunk 6 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:565]: I did think about it, as you recommended it in the linked issue. I am hesitant to introduce more parsing requirements for environment variables if I can avoid it, though it would have been simpler to use that approach in this case (vs. requiring a yaml each loop).

File: eng/common/TestResources/SubConfig-Helpers.ps1 (hunk 3 above)
Comment (weshaggard) [reply eng/common/TestResources/SubConfig-Helpers.ps1:8]: Yes something like that.
